scipy/_lib/doccer.py

```diff
@@ -1,26 +1,39 @@
-''' Utilities to allow inserting docstring fragments for common
-parameters into function and method docstrings'''
+"""Utilities to allow inserting docstring fragments for common
+parameters into function and method docstrings."""
 
+from collections.abc import Callable, Iterable
+from typing import Any, ParamSpec, TypeAlias, TypeVar
 import sys
 
 __all__ = [
-    'docformat', 'inherit_docstring_from', 'indentcount_lines',
-    'filldoc', 'unindent_dict', 'unindent_string', 'extend_notes_in_docstring',
-    'replace_notes_in_docstring', 'doc_replace'
+    "docformat",
+    "inherit_docstring_from",
+    "indentcount_lines",
+    "filldoc",
+    "unindent_dict",
+    "unindent_string",
+    "extend_notes_in_docstring",
+    "replace_notes_in_docstring",
+    "doc_replace",
 ]
 
+P = ParamSpec("P")
+R = TypeVar("R")
 
-def docformat(docstring, docdict=None):
-    ''' Fill a function docstring from variables in dictionary
+Decorator: TypeAlias = Callable[[Callable[..., R]], Callable[..., R]]
+
+
+def docformat(docstring: str, docdict: dict[str, str] | None = None) -> str:
```

Comment:
In case you also want to accept things like `TypedDict` instances, or immutable dict-likes, you could use `Mapping[str, str]` from `collections.abc` instead.